---
title: "VMR 5 - Characterization analyses"
author: "Erick I. Navarro-Delgado"
date: '2023-12-05'
output: 
  html_document:
    toc: true
    toc_float: true
    keep_md: true
---

This document is part of a series of scripts used to estimate the gene-environment contribution to DNAme variability in the CHILD data set. Here, we describe the characterization of the VMRs (in general), and the VMRs per winning model (G, E, G+E and GxE). To do so, we use are using the extended EPIC manifest generated by Dr. Nicole Gladish in 2021, where she annotates the chromatin state of the EPIC probes in cord blood.

```{r}
# Load packages and data
library(circlize)
library(tidyverse)
library(data.table)
library(here)
library(cowplot)
library(clusterProfiler)

VMRs_df = fread(here("Objects/cis_snps_VMRdf.csv"),data.table = FALSE) %>% 
  mutate(probes = str_split(probes, pattern = "\\|"),
         SNP =str_split(SNP, pattern = "\\|"),
         VMR_index = as.character(VMR_index)) 

lmGE_res = fread(here("Objects/lmGE_res_annot_child.csv"), data.table = FALSE) %>% 
  mutate(variables = str_split(variables, pattern = "\\|"),
         VMR_index = as.character(VMR_index))
  
variance_adjusted = readRDS("Objects/variance_adjusted_probes.RDS")

load("EPIC_fdat.RData")  #Full manifest
load("EPIC_Anno_ChromeHMM.RData") #Annotated manifest. 977 CpGs are not present here for some reason compared to the full manifest. I guess it is because those sites did not have an HMM annotation??

# Remove CpG sites from the manifest that were not included in the analysis 
EPIC_Anno_ChromeHMM = EPIC_Anno_ChromeHMM %>% 
  filter(Name %in% row.names(variance_adjusted))

fData_EPIC = fData_EPIC |> 
  filter(TargetID %in% row.names(variance_adjusted))

x = rownames(variance_adjusted)[variance_adjusted %in% EPIC_Anno_ChromeHMM$Name]

x = EPIC_Anno_ChromeHMM %>% 
  filter(!Name %in% row.names(variance_adjusted))
```


# VMR characterization

## Distribution and composition

First, I want to characterize the VMRs's distribution across the human genome. I will also add the CpG probes in the EPIC array as information of the backgrounds probes.

```{r}
# Create data frame with information 
bed_epic = fData_EPIC |> 
  select(c(TargetID, CHR, MAPINFO)) |> 
  rename(chr = CHR,
         probe = TargetID) |> 
  mutate(chr = paste("chr", chr, sep = ""),
         start = MAPINFO,
         end = MAPINFO) |> 
  right_join(variance_adjusted |> 
              rownames_to_column(var = "probe"),
             by = "probe") |> 
  select(c(chr, start, end, probe, variance))

VMRs_bed = VMRs_df |> 
  mutate(seqnames = paste("chr", seqnames, sep = ""),
         VMR = 1) |> 
  rename(chr = seqnames) |> 
  select(c(chr,start,end, VMR))

png(here("Objects/images/distribution_EPICprobes_VMRs.png"),
    units="in", width=4, height=3, res=300)
circos.par("start.degree" = 90, # Start chromosome 1 at 12 o clock
           "gap.degree" = rep(2, 22)) #Increase gaps between chromosomes 
circos.initializeWithIdeogram(chromosome.index = paste0("chr", c(1:22)), #Start ideogram with chromosomes 1-22
                              plotType = c("labels", "ideogram")) 
circos.genomicDensity(VMRs_bed, col ="#1c246c", count_by = "number", border = "#1c246c" )
circos.genomicDensity(bed_epic , col = "#f5ab87", count_by = "number")
circos.clear()
dev.off()

```

Now I want to get information about how many canonical VMRs we have, how many probes they have on average, and how big they are on average. 

```{r}
#Get proportion of canonical and non canonical VMRs
VMRs_df |> 
  mutate(Category = case_when(n_VMPs == 1 ~ "Non-canonical",
                              TRUE ~ "Canonical")) |> 
  group_by(Category) |> 
  summarise(counts = n())
# This info was used for creating a pie chart in excel

#Create VMR summary table
VMRs_df |> 
  filter(n_VMPs > 1) |> 
  select(VMR_index, width, n_VMPs) |> 
  pivot_longer(-VMR_index) |> 
  group_by(name) |> 
  summarise(mean = mean(value), 
            sd = sd(value), 
            min = min(value),
            max = max(value))
```

## Enrichment analyses

I want to characterize now the composition of the probes in the VMRs in terms of their HMM state and their CpG annotation. To do so, I will use the information from Illumina's [manifest](https://support.illumina.com/content/dam/illumina-support/documents/downloads/productfiles/methylationepic/infinium-methylationepic-manifest-column-headings.pdf). The main columns that I am interested in using for the characterization are the following ones: 

**Relation_to_UCSC_CpG_Island**: The location of the CpG relative to the CpG island. 
  -   Shore = 0–2 kb from island.
  -   Shelf = 2–4 kb from island.
  -   N = upstream (5’) of CpG island.
  -   S = downstream (3’) of CpG island.

**UCSC_RefGene_Group**: Gene region feature category describing the CpG position, from UCSC. Features listed in the same order as the target gene transcripts.
  -   TSS200 = 0–200 bases upstream of the transcriptional start site (TSS).
  -   TSS1500 = 200–1500 bases upstream of the TSS.
  -   5'UTR = Within the 5' untranslated region, between the TSS and the ATG start site.
  -   Body = Between the ATG and stop codon; irrespective of the presence of introns, exons, TSS, or promoters.
  -   3'UTR = Between the stop codon and poly A signal.
  
In addition to the default annotation provided by Illumina, Dr Gladish added the expanded 18-chromatin states of several blood cell types from the [Roadmap epigenomics project](https://egg2.wustl.edu/roadmap/web_portal/chr_state_learning.html) produced with HMM. For **cord blood cell types**, she added the **Core 15-state model**. The script used to generate this manifest can be found in /mnt/scratch/KoborLab/shared_coding_resource/ChromeHMM_Data/ChromeHMM_EPIC_Anno_Production.html

Relevant to this project, the Roadmap epigenomics project has chromatin states for the samples E033 (Tcells Cord Blood) and E031 (Bcells Cord Blood). I have to choose one as reference, and I decided to use E033 since the proportion of T cells in my samples are higher compared to B cells (see VMR1 report). This annotations come from the E033_15_coreMarks_dense.bed.gz file obtained from [here](https://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/coreMarks/jointModel/final/). 

More information about this core 15-state model from the Roadmap project website:

> A ChromHMM model applicable to all 127 epigenomes was learned by virtually concatenating consolidated data corresponding to the core set of 5 chromatin marks assayed in all epigenomes (H3K4me3, H3K4me1, H3K36me3, H3K27me3, H3K9me3). The model was trained on 60 epigenomes with highest-quality data (Fig. 2k), which provided sufficient coverage of the different lineages and tissue types (Table S1 - Sheet QCSummary). The ChromHMM parameters used were as follows: Reads were shifted in the 5 to 3 direction by 100 bp. For each consolidated ChIP-seq dataset, read counts were computed in non-overlapping 200 bp bins across the entire genome. Each bin was discretized into two levels, 1 indicating enrichment and 0 indicating no enrichment. The binarization was performed by comparing ChIP-seq read counts to corresponding whole-cell extract control read counts within each bin and using a Poisson p-value threshold of 1e-4 (the default discretization threshold in ChromHMM). We trained several models in parallel mode with the number of states ranging from 10 states to 25 states. We decided to use a 15-state model (Fig. 4a-f, Extended Data 2b) for all further analyses since it captured all the key interactions between the chromatin marks, and because larger numbers of states did not capture sufficiently distinct interactions. The trained model was then used to compute the posterior probability of each state for each genomic bin in each reference epigenome. The regions were labeled using the state with the maximum posterior probability.

### Create background probes and needed objects

```{r}
# Create annotated data set of the probes in VMRs
probes_VMRs = VMRs_df |> 
  mutate(category_vmr = case_when(n_VMPs == 1 ~ "Non-canonical",
                              TRUE ~ "Canonical")) |> 
  dplyr::select(probes, VMR_index, category_vmr) |> 
  unnest(probes) |> 
  left_join(fData_EPIC |> 
              select(TargetID, RELATION_TO_UCSC_CPG_ISLAND) |> 
              dplyr::rename(probes = TargetID,
                     cpg_annotation = RELATION_TO_UCSC_CPG_ISLAND),
            by = "probes") |> 
  mutate(cpg_annotation = case_when(cpg_annotation == "" ~ "Open Sea", #Assign empty cells to Open Sea
                                    TRUE ~ cpg_annotation)) |> 
  left_join(EPIC_Anno_ChromeHMM |> 
              dplyr::select(Name, E033_Tcell_CordBlood_15_State_State) |> 
              dplyr::rename(probes = Name,
                     chrom_state = E033_Tcell_CordBlood_15_State_State),
            by = "probes") |> 
  mutate(chrom_state = str_replace(chrom_state, ".*_", ""),
         chrom_state = case_when(is.na(chrom_state) ~ "Uk",
                                 TRUE ~ chrom_state))

head(probes_VMRs)

# Create background CpG annotation 
probes_background_epic = fData_EPIC |> 
  select(TargetID, RELATION_TO_UCSC_CPG_ISLAND) |> 
  dplyr::rename(probes = TargetID,
         cpg_annotation = RELATION_TO_UCSC_CPG_ISLAND) |> 
  mutate(cpg_annotation = case_when(cpg_annotation == "" ~ "Open Sea", #Assign empty cells to Open Sea
                                    TRUE ~ cpg_annotation)) |> 
  left_join(EPIC_Anno_ChromeHMM |> 
              select(Name, E033_Tcell_CordBlood_15_State_State) |> 
              dplyr::rename(probes = Name,
                     chrom_state = E033_Tcell_CordBlood_15_State_State),
            by = "probes") |> 
  mutate(chrom_state = str_replace(chrom_state, ".*_", ""),
         chrom_state = case_when(is.na(chrom_state) ~ "Uk",
                                 TRUE ~ chrom_state))

head(probes_background_epic)
```

### Enrichment analyses

I will use the clusterProfiler::enricher function, which conducts hypergeometric tests with p adjustment. 

#### CpG Annotation

```{r}
TERM2GENE_cpgannot = probes_background_epic |> 
  select(cpg_annotation, probes)

enrichment_cpgannot_res = clusterProfiler::enricher(probes_VMRs$probes, TERM2GENE = TERM2GENE_cpgannot, maxGSSize = Inf, pAdjustMethod = "fdr")
enrichment_cpgannot_res@result[,c(1:7,9)] 
```

We can observe that only the Open Sea category is over represented (0.67 compared to 0.57 in the background probes)

Now I will plot the proportions

```{r}
# Get significant terms 
significant = enrichment_cpgannot_res@result[,c(1:7,9)] |> 
  filter(p.adjust < 0.05) |> 
  pull(ID)

#Create data frame to plot
cpg_annot_df_plot = TERM2GENE_cpgannot |>
  filter(probes %in% probes_VMRs$probes) |> 
  group_by(cpg_annotation) |> 
  summarise(probes = n()) |> 
  mutate(proportion = probes/length(probes_VMRs$probes)) |> #They sum up to one cause each probe is mapped to only one annotation 
  mutate(cpg_annotation = fct_reorder(cpg_annotation, proportion)) |> 
  arrange(cpg_annotation)

# Create the borders object
signif = enrichment_cpgannot_res@result[,c(1:7,9)] |> 
  mutate(signif = case_when(ID %in% significant ~ "*",
            TRUE ~ NA))
signif = signif[levels(cpg_annot_df_plot$cpg_annotation),]

# Plot

enrichment_annot = cpg_annot_df_plot |> 
  ggplot(aes(x = cpg_annotation, y = proportion)) +
  theme_cowplot() +
  geom_col(fill = "#1c246c") +
  geom_text(aes(label = signif$signif), size = 8, hjust = 0, vjust = 0.8) +
  xlab("CpG annotation") +
  scale_y_continuous(limits = c(0,0.8), breaks = c(0.2, 0.6)) +
  coord_flip()

print(enrichment_annot)

#Same but now save it 
png(here("Objects/images/enrichment_VMRs_cpg_annot.png"),
    units="in", width=2, height=3, res=300)
print(enrichment_annot)
dev.off()

```


#### Chromatin states

```{r}
TERM2GENE_chrom = probes_background_epic |> 
  select(chrom_state, probes)

enrichment_chrom_res = clusterProfiler::enricher(probes_VMRs$probes, TERM2GENE = TERM2GENE_chrom, maxGSSize = Inf, pAdjustMethod = "fdr")
enrichment_chrom_res@result[,c(1:7,9)] 

significant_chrom = enrichment_chrom_res@result[,c(1:7,9)] |> 
  filter(p.adjust < 0.05) |> 
  pull(ID)

#Create data frame to plot
chrom_df_plot = TERM2GENE_chrom |>
  filter(probes %in% probes_VMRs$probes) |> 
  group_by(chrom_state) |> 
  summarise(probes = n()) |> 
  mutate(proportion = probes/length(probes_VMRs$probes)) |> #They sum up to one cause each probe is mapped to only one annotation 
  mutate(chrom_state = fct_reorder(chrom_state, proportion)) |> 
  arrange(chrom_state) |> 
  slice_tail(n = 10) |> # Use only the top 10
  droplevels()

# Create the borders object
signif_chrom = enrichment_chrom_res@result[,c(1:7,9)] |> 
  rowwise() |> 
  mutate(gene_ratio_solved = eval(parse(text = GeneRatio)),
         bg_ratio_solved = eval(parse(text = BgRatio)),
         OR = gene_ratio_solved / bg_ratio_solved ) |> 
  mutate(signif = case_when((ID %in% significant_chrom) & (OR > 1) ~ "*",
                             (ID %in% significant_chrom) & (OR < 1) ~ "#8598ba", #Unnecesary lol, it is an enrichment analysis, which is a one sided fisher's test - so no depletions are found here. 
                             TRUE ~ NA)) |> 
  column_to_rownames(var = "ID")
signif_chrom = signif_chrom[levels(chrom_df_plot$chrom_state),]

# Plot
enrichment_chrom = chrom_df_plot |> 
  ggplot(aes(x = chrom_state, y = proportion)) +
  geom_col(fill = "#1c246c") +
  geom_text(aes(label = signif_chrom$signif), size = 8, hjust = 0, vjust = 0.8) +
  xlab("CpG annotation") +
  theme_cowplot() +
  scale_y_continuous(limits = c(0, 0.45), breaks = c(0.1, 0.3)) +
  coord_flip()

print(enrichment_chrom)

png(here("Objects/images/enrichment_VMRs_chrom.png"),
    units="in", width=2, height=3, res=300)
print(enrichment_chrom)
dev.off()

```

We can see that VMRs are enriched in Quies, ReprPCWk, Enh, Het and ReprPC. And that most VMRs are in quiescent state. 

#### Gene context

I will conduct the same descriptive analysis but with gene context as categories. Here I will have to create a different background data set, since each CpG site can belong to one or more categories (e.g. it can be in the body of one gene and in the promoter of another one at the same time). 

```{r}
TERM2GENE_gene_context = fData_EPIC |> 
  select(TargetID, UCSC_REFGENE_GROUP) |> 
  mutate(UCSC_REFGENE_GROUP = case_when(UCSC_REFGENE_GROUP == "" ~ "Intergenic", #Set empty categories to intergenic
                                        TRUE ~ UCSC_REFGENE_GROUP),
         UCSC_REFGENE_GROUP = str_split(UCSC_REFGENE_GROUP, pattern = ";", simplify = FALSE)) |> #Separate CpGs that have multiple annotations
  unnest(UCSC_REFGENE_GROUP) |> # Convert it to a longer data frame
  unique() |>  #Remove repetitive entries (e.g. a CpG being in body of 2 different genes)
  rename(probes = TargetID, gene_context = UCSC_REFGENE_GROUP) |> #Rename
  select(gene_context, probes) #Set the correct order


enrichment_genecont_res = clusterProfiler::enricher(probes_VMRs$probes, TERM2GENE = TERM2GENE_gene_context, maxGSSize = Inf, pAdjustMethod = "fdr")
enrichment_genecont_res@result[,c(1:7,9)] 

significant_genecont = enrichment_genecont_res@result[,c(1:7,9)] |> 
  filter(p.adjust < 0.05) |> 
  pull(ID)

#Create data frame to plot
genecont_df_plot = TERM2GENE_gene_context |>
  filter(probes %in% probes_VMRs$probes) |> 
  group_by(gene_context) |> 
  summarise(probes = n()) |> 
  mutate(proportion = probes/length(probes_VMRs$probes)) |> #They sum up to one cause each probe is mapped to only one annotation 
  mutate(gene_context = fct_reorder(gene_context, proportion)) |> 
  arrange(gene_context)

# Create the borders object
signif_genecont = enrichment_genecont_res@result[,c(1:7,9)] |> 
  mutate(signif = case_when(ID %in% significant_genecont ~ "*",
                             TRUE ~ NA)) 
signif_genecont = signif_genecont[levels(genecont_df_plot$gene_context),]

# Plot
enrichment_genecont = genecont_df_plot |> 
  ggplot(aes(x = gene_context, y = proportion)) +
  theme_cowplot() +
  geom_col(fill = "#1c246c") +
  geom_text(aes(label = signif_genecont$signif), size = 8, hjust = 0, vjust = 0.8) +
  xlab("CpG annotation") +
  scale_y_continuous(limits = c(0,0.45), breaks = c(0.2, 0.4)) +
  coord_flip()

print(enrichment_genecont)

png(here("Objects/images/enrichment_VMRs_genecont.png"),
    units="in", width=2, height=3, res=300)
print(enrichment_genecont)
dev.off()
```

We can see here that VMRs are most usually in body and intergenic regions, and they are enriched in intergenic probes in comparison to the background probes in the EPIC array.

# Winning model group characterization 

To characterize the winning models, I will conduct enrichment analyses **on the VMRs** (not on the probes as I did for characterizing the VMRs). As background, I will use the VMRs, and I will do enrichment analyses on each winning model group (G, E, G+E and GxE). 

## Enrichment analysis 

### CpG annotation
```{r}
#Get the background object
TERM2_VMR_cpgann = VMRs_df |> 
  unnest(probes) |> 
  left_join(TERM2GENE_cpgannot, by = "probes") |> # Add info about the cpg annotation of the probes in each vmr
  select(cpg_annotation, VMR_index) |> 
  unique() 

VMR_clusters = list(
  "G" = lmGE_res |>
    filter(model_group == "G") |> 
    pull(VMR_index),
  "E" = lmGE_res |>
    filter(model_group == "E") |> 
    pull(VMR_index),
  "G+E" = lmGE_res |>
    filter(model_group == "G+E") |> 
    pull(VMR_index),
  "GxE" = lmGE_res |>
    filter(model_group == "GxE") |> 
    pull(VMR_index)
)
  

enrich_vmrs_cpgannot <- compareCluster(VMR_clusters, enricher, TERM2GENE = TERM2_VMR_cpgann, pAdjustMethod = "fdr", minGSSize = 0, maxGSSize = Inf, pvalueCutoff  = 1, qvalueCutoff = 1)

```


### Chromatin state

```{r}
TERM2_VMR_chrom= VMRs_df |> 
  unnest(probes) |> 
  left_join(TERM2GENE_chrom, by = "probes") |> # Add info about the cpg annotation of the probes in each vmr
  select(chrom_state, VMR_index) |> 
  unique() 

enrich_vmrs_chrom <- compareCluster(VMR_clusters, enricher, TERM2GENE = TERM2_VMR_chrom, pAdjustMethod = "fdr", minGSSize = 0, maxGSSize = Inf, pvalueCutoff  = 1, qvalueCutoff = 1)

```

### Gene context 

```{r}
TERM2_VMR_genecont = VMRs_df |> 
  unnest(probes) |> 
  left_join(TERM2GENE_gene_context, by = "probes", multiple = "all") |> # Add info about the cpg annotation of the probes in each vmr
  select(gene_context, VMR_index) |> 
  unique() 

enrich_vmrs_genecont <- compareCluster(VMR_clusters, 
                                       enricher, TERM2GENE = TERM2_VMR_genecont, pAdjustMethod = "fdr", minGSSize = 0, maxGSSize = Inf, pvalueCutoff  = 1, qvalueCutoff = 1)

```

### Make a single plot 

The dot plots above are not nice toegether (some display only 2/4 categories, size scale is different, etc.). So, I will gather the results and create a single plot

```{r}
VMR_enrich_plot = enrich_vmrs_genecont@compareClusterResult |> 
  select(Cluster, ID, Count, p.adjust, GeneRatio) |> 
  mutate(Comparison = "Gene context") |> 
  bind_rows(enrich_vmrs_chrom@compareClusterResult |> 
              select(Cluster, ID, Count, p.adjust) |> 
              mutate(Comparison = "Chromatin state") |> 
              filter(ID != "Uk")) |> #Remove the VMRs with unknown state 
  bind_rows(enrich_vmrs_cpgannot@compareClusterResult |> 
              select(Cluster, ID, Count, p.adjust) |>
              mutate(Comparison = "CpG Island")) |> 
  mutate(FDR = case_when(p.adjust < 0.05 ~ "< 0.05", 
                                 TRUE ~ ">= 0.05"),
         ID = fct_reorder(ID, Count)) |> #Reorder order of appearance in the plot
  mutate(Cluster = factor(Cluster, levels = c("E", "G", "G+E", "GxE"))) |> 
  ggplot(aes(x = Cluster, y = ID, size = Count, color = FDR)) +
  geom_point() +
  facet_grid("Comparison", 
             scales = "free", # Categories only appear in their enrichment group
             space = "free_y") + # Different grids will have different space in the plot 
  DOSE::theme_dose(font.size = 12) + 
  xlab("") +
  ylab("") + 
  scale_color_manual(values = c(">= 0.05" = "grey88" ,"< 0.05" = "tomato"))

print(VMR_enrich_plot)

png(here("Objects/images/enrichment_VMR_models.png"),
    units="in", width=4.3, height=6.5, res=300)
print(VMR_enrich_plot)
dev.off()
```

